src/core/thread/network_data_publisher.cpp

```diff
@@ -0,0 +1,601 @@
+        "Added",    // (3) kAdded
+        "Removing", // (4) kRemoving
+    };
+
+    static_assert(0 == kNoEntry, "kNoEntry value is not correct");
+    static_assert(1 == kToAdd, "kToAdd value is not correct");
+    static_assert(2 == kAdding, "kAdding value is not correct");
+    static_assert(3 == kAdded, "kAdded value is not correct");
+    static_assert(4 == kRemoving, "kRemoving value is not correct");
+
+    return kStateStrings[aState];
+}
+
+#if OPENTHREAD_CONFIG_TMF_NETDATA_SERVICE_ENABLE
+
+//---------------------------------------------------------------------------------------------------------------------
+// Publisher::DnsSrpServiceEntry
+
+Publisher::DnsSrpServiceEntry::DnsSrpServiceEntry(Instance &aInstance)
+    : Entry(aInstance)
+{
+}
+
+void Publisher::DnsSrpServiceEntry::PublishAnycast(uint8_t aSequenceNumber)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service anycast (seq-num:%d)", aSequenceNumber);
+
+    if (GetState() != kNoEntry)
+    {
+        if ((mType == kTypeAnycast) && (mInfo.mAnycast.mSequenceNumber == aSequenceNumber))
+        {
+            otLogInfoNetData("Publisher: %s is already being published", ToString().AsCString());
+            ExitNow();
+        }
+
+        Remove(/* aRegisterWithLeader */ true);
+    }
+
+    mType                          = kTypeAnycast;
+    mInfo.mAnycast.mSequenceNumber = aSequenceNumber;
+    SetState(kToAdd);
+
+    Process();
+
+exit:
+    return;
+}
+
+void Publisher::DnsSrpServiceEntry::PublishUnicast(const Ip6::Address &aAddress, uint16_t aPort)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service unicast (%s, port:%d)", aAddress.ToString().AsCString(),
+                     aPort);
+
+    if (GetState() != kNoEntry)
+    {
+        if ((mType == kTypeUnicast) && (mInfo.mUnicast.mAddress == aAddress) && (mInfo.mUnicast.mPort == aPort))
+        {
+            otLogInfoNetData("Publisher: %s is already being published", ToString().AsCString());
+            ExitNow();
+        }
+
+        Remove(/* aRegisterWithLeader */ true);
+    }
+
+    mType                   = kTypeUnicast;
+    mInfo.mUnicast.mAddress = aAddress;
+    mInfo.mUnicast.mPort    = aPort;
+    SetState(kToAdd);
+
+    Process();
+
+exit:
+    return;
+}
+
+void Publisher::DnsSrpServiceEntry::PublishUnicast(uint16_t aPort)
+{
+    otLogInfoNetData("Publisher: Publishing DNS/SRP service unicast (ml-eid, port:%d)", aPort);
+
+    if (GetState() != kNoEntry)
+    {
+        if ((mType == kTypeUincastMeshLocalEid) && (mInfo.mUnicast.mPort == aPort))
+        {
+            otLogInfoNetData("Publisher: %s is already being published", ToString().AsCString());
+            ExitNow();
+        }
+
+        Remove(/* aRegisterWithLeader */ true);
+    }
+
+    mType                   = kTypeUincastMeshLocalEid;
+    mInfo.mUnicast.mAddress = Get<Mle::Mle>().GetMeshLocal64();
+    mInfo.mUnicast.mPort    = aPort;
+    SetState(kToAdd);
+
+    Process();
```

Comment:
This code seems to be repeated at least three times - is it beneficial to break into a separate private method?

Reply:
Good suggestion. Thanks. I need to think if/how we can refactor this (since the code here is somewhat tied to each `Type` and its associated parameters).